rust/noosphere-storage/src/implementation/rocks_db.rs

```diff
@@ -0,0 +1,218 @@
+            cfs.push(ColumnFamilyDescriptor::new(*store_name, cf_opts));
+        }
+
+        let mut db_opts = Options::default();
+        db_opts.create_if_missing(true);
+        db_opts.create_missing_column_families(true);
+
+        Ok(DbInner::open_cf_descriptors(&db_opts, path, cfs)?)
+    }
+}
+
+#[async_trait]
+impl Storage for RocksDbStorage {
+    type BlockStore = RocksDbStore;
+    type KeyValueStore = RocksDbStore;
+
+    async fn get_block_store(&self, name: &str) -> Result<Self::BlockStore> {
+        self.get_store(name).await
+    }
+
+    async fn get_key_value_store(&self, name: &str) -> Result<Self::KeyValueStore> {
+        self.get_store(name).await
+    }
+}
+
+impl Drop for RocksDbStorage {
+    fn drop(&mut self) {
+        // Clear out the weak ref if this is the only/last reference.
+        if let Ok(mut db_handles) = DB_HANDLES.lock() {
+            if let Some(weak_ref) = db_handles.get(&self.path) {
+                if weak_ref.strong_count() == 1 {
+                    db_handles.remove(&self.path);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct RocksDbStore {
+    name: String,
+    db: Arc<DbInner>,
+}
+
+impl RocksDbStore {
+    pub fn new(db: Arc<DbInner>, name: String) -> Result<Self> {
+        Ok(RocksDbStore { db, name })
+    }
+
+    /// Returns the column family handle. Unfortunately generated on every call
+    /// due to not being `Sync`, potentially `unsafe` alternatives:
+    /// https://github.com/rust-rocksdb/rust-rocksdb/issues/407
+    fn cf_handle<'a>(&'a self) -> Result<ColumnType> {
+        self.db
+            .cf_handle(&self.name)
+            .ok_or_else(move || anyhow!("Could not open handle for {}", self.name))
+    }
+}
+
+#[async_trait]
+impl Store for RocksDbStore {
+    async fn read(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let cf = self.cf_handle()?;
+        #[cfg(feature = "rocksdb-mt")]
+        let cf = &cf;
+        Ok(self.db.get_cf(cf, key)?)
+    }
+
+    async fn write(&mut self, key: &[u8], bytes: &[u8]) -> Result<Option<Vec<u8>>> {
+        let cf = self.cf_handle()?;
+        #[cfg(feature = "rocksdb-mt")]
+        let cf = &cf;
+        let old_bytes = self.db.get_cf(cf, key)?;
+        self.db.put_cf(cf, key, bytes)?;
+        Ok(old_bytes)
+    }
+
+    async fn remove(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let cf = self.cf_handle()?;
+        #[cfg(feature = "rocksdb-mt")]
+        let cf = &cf;
+        let old_bytes = self.db.get_cf(cf, key)?;
+        self.db.delete_cf(cf, key)?;
+        Ok(old_bytes)
+    }
+
+    async fn flush(&self) -> Result<()> {
+        // With the use of WAL, we do not want to actively flush on every sync,
+        // and instead allow RocksDB to determine when to flush to OS.
+
+        // let cf = self.cf_handle()?;
+        // #[cfg(feature = "rocksdb-mt")]
+        // let cf = &cf;
+        // self.db.flush_cf(cf)?;
```

Comment:
Is there a TODO to go with this commented code?

Reply:
Just an example of flushing, we can toss it